### PR TITLE
Scopes in mir

### PR DESCRIPTION
**Pull request by nikomatsakis**
This PR adds scopes to MIR. There is a tree of scopes (each represented by a `ScopeId`). Every statement, variable, and terminator now has an associated scope and span.  It also adds a `-Z dump-mir` switch one can use to conveniently examine the MIR as optimizations proceed.

The intention is two-fold. First, to support MIR debug-info. This PR does not attempt to modify trans to make use of the scope information, however.

Second, in a more temporary capacity, to support the goal of moving regionck and borowck into the MIR. To that end, the PR also constructs a "scope auxiliary" table storing the extent of each span (this is kept separate from the main MIR, since it contains node-ids) and the dom/post-dom of the region in the graph where the scope occurs. When we move to non-lexical lifetimes, I expect this auxiliary information to be discarded, but that is still some ways in the future (requires, at minimum, an RFC, and there are some thorny details to work out -- though I've got an in-progress draft).

Right now, I'm just dropping this auxiliary information after it is constructed. I was debating for some time whether to add some sort of sanity tests, but decided to just open this PR instead, because I couldn't figure out what such a test would look like (and we don't have independent tests for this today beyond the regionck and borrowck tests).

I'd prefer not to store the auxiliary data into any kind of "per-fn" map. Rather, I'd prefer that we do regionck/borrowck/whatever-else immediately after construction -- that is, we build the MIR for fn X and immediately thereafter do extended correctness checking on it. This will reduce peak memory usage and also ensure that the auxiliary data doesn't exist once optimizations begin. It also clarifies the transition point where static checks are complete and MIR can be more freely optimized.

cc @rust-lang/compiler @nagisa

### CODE REVIEW

File: src/librustc/mir/repr.rs [hunk 1]
```diff
@@ -614,12 +618,60 @@ impl<'tcx> Debug for Lvalue<'tcx> {
...
+
+impl IndexMut<ScopeId> for ScopeDataVec {
+    #[inline]
+    fn index_mut(&mut self, index: ScopeId) -> &mut ScopeData {
+        &mut self.vec[index.index()]
+    }
+}
+
+#[derive(Copy, Clone, Debug, RustcEncodable, RustcDecodable)]
+pub struct ScopeId(u32);
```
Comment (eddyb) [inline src/librustc/mir/repr.rs]: I *almost* feel like we should have a general-purpose `Id<T>` type for these kinds of things.

File: src/librustc_mir/build/mod.rs [hunk 2]
```diff
@@ -33,6 +51,42 @@ struct CFG<'tcx> {
...
+pub struct Location {
+    /// the location is within this block
+    pub block: BasicBlock,
+
+    /// the location is the start of the this statement; or, if `statement_index`
+    /// == num-statements, then the start of the terminator.
+    pub statement_index: usize,
+}
+
+pub struct MirPlusPlus<'tcx> {
```
Comment (eddyb) [inline src/librustc_mir/build/mod.rs]: What about something along the lines of `MirAndRegionScopes`?

File: src/librustc/mir/repr.rs [hunk 3]
```diff
@@ -228,8 +228,13 @@ pub struct BasicBlockData<'tcx> {
...
 }
 
+#[derive(Clone, Debug, RustcEncodable, RustcDecodable)]
+pub struct Terminator<'tcx> {
+    pub kind: TerminatorKind<'tcx>
+}
+
 #[derive(Clone, RustcEncodable, RustcDecodable)]
-pub enum Terminator<'tcx> {
+pub enum TerminatorKind<'tcx> {
```
Comment (eddyb) [inline src/librustc/mir/repr.rs]: *In a galaxy far, far away* they have neater ADTs :disappointed:

File: src/librustc/mir/repr.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/librustc/mir/repr.rs]: I've been wanting for some time to make a macro for this purpose.

Comment (eddyb) [inline src/librustc/mir/repr.rs]: At least in the HIR, it would be more practical to use a single generic type IMO.

File: src/librustc/mir/repr.rs [hunk 4]
```diff
@@ -155,6 +159,8 @@ pub struct VarDecl<'tcx> {
     pub mutability: Mutability,
     pub name: Name,
     pub ty: Ty<'tcx>,
+    pub scope: ScopeId, // scope in which variable was declared
```
Comment (nagisa) [inline src/librustc/mir/repr.rs]: NIT: use doc-comments to document fields.

File: src/librustc/mir/repr.rs [hunk 5]
```diff
@@ -32,6 +32,10 @@ pub struct Mir<'tcx> {
     /// that indexes into this vector.
     pub basic_blocks: Vec<BasicBlockData<'tcx>>,
 
+    /// List of lexical scopes; these are referenced by statements and
+    /// used (eventually) for debuginfo. Indexed by a `ScopeId`.
+    pub scopes: ScopeDataVec,
```
Comment (nagisa) [inline src/librustc/mir/repr.rs]: I feel like it might be a good idea to have spans in the side table, along with scopes.

File: src/librustc/mir/repr.rs [hunk 6]
```diff
@@ -614,12 +618,60 @@ impl<'tcx> Debug for Lvalue<'tcx> {
 }
 
 ///////////////////////////////////////////////////////////////////////////
+// Scopes
+
+#[derive(Clone, Debug, RustcEncodable, RustcDecodable)]
+pub struct ScopeDataVec {
```
Comment (nagisa) [inline src/librustc/mir/repr.rs]: Why the newtype?

File: src/librustc_mir/build/mod.rs [hunk 7]
```diff
@@ -21,8 +21,26 @@ use syntax::codemap::Span;
 pub struct Builder<'a, 'tcx: 'a> {
     hir: Cx<'a, 'tcx>,
     cfg: CFG<'tcx>,
+
+    // the current set of scopes, updated as we traverse;
+    // see the `scope` module for more details
```
Comment (nagisa) [inline src/librustc_mir/build/mod.rs]: NIT: doc-comment

File: src/librustc_mir/build/mod.rs (hunk 2 above)
Comment (nagisa) [inline src/librustc_mir/build/mod.rs]: Why not just return `(Mir<'tcx>, Vec<ScopeAuxiliary>)` from the construct instead?

File: src/librustc_mir/build/scope.rs [hunk 8]
```diff
@@ -98,8 +98,11 @@ use rustc::middle::const_eval::ConstVal;
 use rustc_const_eval::ConstInt;
 
 pub struct Scope<'tcx> {
+    // the scope-id within the scope_data_vec
```
Comment (nagisa) [inline src/librustc_mir/build/scope.rs]: Nit: doc.

File: src/librustc_mir/build/scope.rs [hunk 9]
```diff
@@ -219,26 +222,44 @@ impl<'a,'tcx> Builder<'a,'tcx> {
...
         // to make sure all the `cached_block`s are filled in.
         self.diverge_cleanup();
         let scope = self.scopes.pop().unwrap();
         assert_eq!(scope.extent, extent);
-        build_scope_drops(&mut self.cfg, &scope, &self.scopes[..], block)
+        unpack!(block = build_scope_drops(&mut self.cfg, &scope, &self.scopes, block));
+        self.scope_auxiliary[scope.id.index()]
+            .postdoms
+            .push(self.cfg.current_location(block));
+        block.and(())
```
Comment (nagisa) [inline src/librustc_mir/build/scope.rs]: block.unit()

File: src/librustc_mir/mir_map.rs [hunk 10]
```diff
@@ -182,8 +182,14 @@ fn build_mir<'a,'tcx:'a>(cx: Cx<'a,'tcx>,
     let parameter_scope =
         cx.tcx().region_maps.lookup_code_extent(
             CodeExtentData::ParameterScope { fn_id: fn_id, body_id: body.id });
-    let mut mir = build::construct(cx, span, implicit_arg_tys, arguments,
-                                  parameter_scope, fn_sig.output, body);
+    let MirPlusPlus { mut mir, scope_auxiliary: _ } =
```
Comment (nagisa) [inline src/librustc_mir/mir_map.rs]: Seriously, perhaps use a tuple?

File: src/librustc_mir/build/matches/mod.rs [hunk 11]
```diff
@@ -106,7 +105,7 @@ impl<'a,'tcx> Builder<'a,'tcx> {
 
     pub fn expr_into_pattern(&mut self,
                              mut block: BasicBlock,
-                             var_extent: CodeExtent, // lifetime of vars
+                             var_scope_id: ScopeId, // lifetime of vars
```
Comment (nagisa) [inline src/librustc_mir/build/matches/mod.rs]: This technically isn’t lifetime of the variable anymore, so the comment should be adjusted.

File: src/librustc_trans/trans/mir/block.rs [hunk 12]
```diff
@@ -569,3 +569,4 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
         self.blocks[bb.index()].llbb
     }
 }
+
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/block.rs]: Nit: trailing newline

File: src/librustc_mir/build/mod.rs [hunk 13]
```diff
@@ -160,14 +160,17 @@ pub fn construct<'a,'tcx>(hir: Cx<'a,'tcx>,
...
     let mut block = START_BLOCK;
-    let arg_decls = unpack!(block = builder.args_and_body(block,
-                                                          implicit_arguments,
-                                                          explicit_arguments,
-                                                          argument_extent,
-                                                          ast_block));
-
-    builder.cfg.terminate(block, TerminatorKind::Goto { target: END_BLOCK });
-    builder.cfg.terminate(END_BLOCK, TerminatorKind::Return);
+    let (arg_decls, arg_scope_id) =
```
Comment (nagisa) [inline src/librustc_mir/build/mod.rs]: Is the `arg_scope_id` here the ID of the scope of whole function?

File: src/librustc_mir/build/scope.rs [hunk 14]
```diff
@@ -288,14 +289,19 @@ impl<'a,'tcx> Builder<'a,'tcx> {
...
+                self.cfg.terminate(block, scope.id, span, free);
                 block = next;
             }
             self.scope_auxiliary[scope.id.index()]
                 .postdoms
                 .push(self.cfg.current_location(block));
         }
-        self.cfg.terminate(block, TerminatorKind::Goto { target: target });
+
+        let scope_id = self.innermost_scope_id();
```
Comment (nagisa) [inline src/librustc_mir/build/scope.rs]: I suspect this might be not correct. AFAIR the loop above does not pop any scopes, therefore this will be the scope_id of the inner-most scope. I’m not sure if that’s the real intention.

File: src/librustc/mir/repr.rs (hunk 6 above)
Comment (eddyb) [inline src/librustc/mir/repr.rs]: For the `Index` impl, presumably.

File: src/librustc_mir/pretty.rs [hunk 15]
```diff
@@ -16,21 +16,33 @@ use syntax::ast::NodeId;
...
-    for (&nodeid, mir) in iter {
-        try!(write_mir_intro(tcx, nodeid, mir, w));
-        // Nodes
-        for block in mir.all_basic_blocks() {
-            try!(write_basic_block(block, mir, w));
-        }
-        try!(writeln!(w, "}}"))
+pub fn write_mir_pretty<'a, 'tcx, I>(tcx: &ty::TyCtxt<'tcx>,
+                                     iter: I,
+                                     w: &mut Write)
```
Comment (nagisa) [inline src/librustc_mir/pretty.rs]: I’m not opposed to that per se, but any reason to do dynamic dispatch here?

File: src/librustc/session/config.rs [hunk 16]
```diff
@@ -663,6 +663,8 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
           "print the result of the translation item collection pass"),
     mir_opt_level: Option<usize> = (None, parse_opt_uint,
           "set the MIR optimization level (0-3)"),
+    dump_mir: Option<String> = (None, parse_opt_string,
```
Comment (nagisa) [inline src/librustc/session/config.rs]: Is `--unpretty={mir[=function],mir-cfg[=function]` somehow unsatisfactory?

File: src/librustc_mir/transform/simplify_cfg.rs [hunk 17]
```diff
@@ -106,8 +113,11 @@ impl SimplifyCfg {
 
 impl<'tcx> MirPass<'tcx> for SimplifyCfg {
     fn run_pass(&mut self, tcx: &TyCtxt<'tcx>, id: NodeId, mir: &mut Mir<'tcx>) {
+        let mut counter = 0;
         let mut changed = true;
         while changed {
+            pretty::dump_mir(tcx, "simplify_cfg", &counter, id, mir, None);
```
Comment (nagisa) [inline src/librustc_mir/transform/simplify_cfg.rs]: I suppose that’s for debugging purposes?

File: src/librustc_mir/pretty.rs (hunk 15 above)
Comment (eddyb) [inline src/librustc_mir/pretty.rs]: Also, wouldn't `&mut fmt::Formatter` be better?

File: src/librustc/mir/repr.rs (hunk 6 above)
Comment (nagisa) [inline src/librustc/mir/repr.rs]:
```
impl ::std::ops::Index<ScopeId> for Vec<ScopeData> {
    ...
}
```

does not violate any of our current coherence rules.

**Comment by nikomatsakis** [comment]
Haven't had time to review comments yet; I just added a kind of drive-by cleanup of scope dropping code to make it a bit less complex, and supply a better span to the resume terminator. Seems to work locally but I havent' done a `-Z orbit` build.

File: src/librustc/mir/repr.rs (hunk 5 above)
Comment (nikomatsakis) [inline src/librustc/mir/repr.rs]: @nagisa 

> I feel like it might be a good idea to have spans in the side table, along with scopes.

While every reference to a scope has a span, those spans are different at each point, so we can't put them in the same table. It might be useful to put them in different tables. But then I think MIR would be the wrong place to do this -- more like codemap. (And, to some extent, it already does.)

File: src/librustc/session/config.rs (hunk 16 above)
Comment (nikomatsakis) [inline src/librustc/session/config.rs]: @nagisa 

> Is --unpretty={mir[=function],mir-cfg[=function]} somehow unsatisfactory/not enough?

It is indeed unsatisfactory for two reasons:

1. It dumps the IR only after all the work is done, rather than dumping the state in between every pass and during processing. 
2. It offers no filters on the state that is dumped. i.e., usually I'm debugging a problem in some particular function or is some particular pass, and I'd like to be able to just dump the MIR for that.
3. In general `--unpretty` does not fit well with our stabilization scheme. I do not see us *ever* stabilizing MIR dumps -- or at least not for a very long time -- which means that `-Z` is a more appropriate place for the option to live. This has the added benefit that we don't have to write `-Z unstable-options` when using it, which is annoying to do. (If you find yourself writing `-Z unstable-options`, and the flag feels like something internal, it should probably just *be* a `-Z` flag.)

File: src/librustc_mir/transform/simplify_cfg.rs (hunk 17 above)
Comment (nikomatsakis) [inline src/librustc_mir/transform/simplify_cfg.rs]: @nagisa 

> I suppose that was for debugging purposes?

I was debugging things, but I left it in on purpose. My intention is that eventually we will have the option to dump MIR at every potentially interesting point during execution.

File: src/librustc/mir/repr.rs (hunk 6 above)
Comment (nikomatsakis) [inline src/librustc/mir/repr.rs]: > ... does not violate any of our current coherence rules.

I suppose that's true. OK.

Comment (nikomatsakis) [inline src/librustc/mir/repr.rs]: I think my thinking was that I only wanted `ScopeId` to be usable as an index in *this particular* vector, but `Vec<ScopeData>` is probably enough to ensure that.

File: src/librustc_mir/build/scope.rs [hunk 18]
```diff
@@ -602,58 +612,42 @@ fn build_scope_drops<'tcx>(cfg: &mut CFG<'tcx>,
...
+        cached_block
+    } else if earlier_scopes.is_empty() {
+        // Diverging from the root scope creates a RESUME terminator.
+        // FIXME what span to use here?
+        let resumeblk = cfg.start_new_cleanup_block();
+        cfg.terminate(resumeblk, scope.id, DUMMY_SP, TerminatorKind::Resume);
+        resumeblk
+    } else {
+        // Diverging from any other scope chains up to the previous scope.
+        build_diverge_scope(tcx, cfg, unit_temp, earlier_scopes)
```
Comment (nagisa) [inline src/librustc_mir/build/scope.rs]: I originally avoided recursion, because the common case is to have many scopes each having few drops. This means that even for simple functions we would recurse a lot (on top of already big amount of recursion in other MIR build code), which potentially will not work out very well on small stack environments.

File: src/librustc_mir/build/scope.rs [hunk 19]
```diff
@@ -665,29 +659,35 @@ fn build_diverge_scope<'tcx>(tcx: &TyCtxt<'tcx>,
...
-                          value: previous_value,
-                          target: target,
-                          unwind: None
-                      });
-        last_drop_block.unwrap()
-    } else {
-        // If `previous.is_none()`, there were no drops in this scope – we return the
-        // target.
-        target
+    // Next, build up the drops. Here we iterate the vector in
```
Comment (nagisa) [inline src/librustc_mir/build/scope.rs]: Weren’t we supposed to execute drops first and then frees second (drops may depend on memory still being allocated, but frees do not otherwise care about things being dropped)?

File: src/librustc_mir/build/scope.rs (hunk 18 above)
Comment (nagisa) [inline src/librustc_mir/build/scope.rs]: This will also needlessly recurse on empty scopes (which are quite common), which were previously ignored by the `filter(|s| !s.drops.is_empty() || s.free.is_some())`.

**Comment by nikomatsakis** [comment]
@nagisa 

I missed two of your comments, an you raise a good point! You wrote:

> Is the arg_scope_id here the ID of the scope of whole function?

and

> I suspect this might be not correct. AFAIR the loop above does not pop any scopes, therefore this will be the scope_id of the inner-most scope. I’m not sure if that’s the real intention.

The way I coded it up, every terminator etc must have a scope, but that leaves the question of what scope the *return block* itself is. Perhaps we want an outer-most scope that is never popped. This would I guess correspond to the `ParameterScope` extent. Return would target that extent (but not exit it).
Hmm, good catch, I think you are correct that the scope there is probably wrong. I wll ad

**Comment by nikomatsakis** [comment]
er, not the ParameterScope, but the `CallSiteScope`

**Comment by nikomatsakis** [comment]
Rebased, make check passes locally.

**Comment by nikomatsakis** [comment]
@nagisa 

> Weren’t we supposed to execute drops first and then frees second (drops may depend on memory still being allocated, but frees do not otherwise care about things being (not-)dropped)?

I believe the code as written *does* execute drops first. Note that it builds up the drops in reverse.

**Comment by nikomatsakis** [comment]
@nagisa btw, the review sort of "defaulted" to @Aatch, which is fine with me -- but I don't know how much time he has. You want to be reviewer? (seeing as you've already taken a look) Does anyone have objections to this basic approach?

One thing that occurred to me while discussing with @pnkfelix -- we should figure out how we plan to handle "lifetime start/lifetime end" intrinsics in MIR trans. If we want to emit them when we exit a *scope*, we have to keep the "scope-auxiliary" information up-to-date. This seems plausible, but it may be easier to do it by having a `SCOPE_EXIT(x)` *statement* (so it would be carried around automatically). If we added such an instruction, it would also be nice for dataflow, I think, since it would correspond to kills (e.g. for all loans tied to that scope). (In trans, it would correspond to emitting lifetime-end intrinsics).

An alternative would be added lifetime-end explicitly in MIR, but it seems like that will result in many more instructions -- you would need `N*M` "words" where `N` is number of variables and `M` is number of exits from a scope, vs `N+M` (one word per variable to tie it to a scope, one word per exit for the statement).

**Comment by nikomatsakis** [comment]
(To be clear, if we were going to have `SCOPE_EXIT` statements, I'd remove the `doms/postdoms` fields from "scope auxiliary" -- and I guess just change it to `ScopeExtents`)

**Comment by nagisa** [comment]
@nikomatsakis I don’t think I have much more outstanding comments other than my dislike about needlessly making `exit_scope` recursive again.

re lifetimes: it seems to me we already store the necessary information in scopes and could implicitly insert e.g. lifetime-end on terminators having a matching scope. I’m not sure it is necessary to encode MIR variable lifetimes in MIR itself. At very least it would make writing MIR passes considerably more complex, since most of these do not need to care about variable lifetimes at all.

**Comment by nagisa** [comment]
r? @nagisa

Sure.

**Comment by nikomatsakis** [comment]
> re lifetimes: it seems to me we already store the necessary information in scopes and could implicitly insert e.g. lifetime-end on scope exits.

I don't quite follow what you mean. The current version of this branch stores the information in the doms/post-doms vectors, which are going to be a pain to maintain (much harder than pseudo-instructions, which virtually  maintain themselves; even better if they take a vector of operands that link to the variables that are to be dropped, though that gets back to the N*M storage question). I was imagining that we would just throw them way once optimization begins, but of course we can't do that if we need them in mir-trans.

I was thinking that perhaps we could recompute the end-lifetime instructions, but that's not really true. The problem is borrows: we'd need (at least) the region information for each borrow, which I plan to erase. But even if we didn't erase it, those are just going to be regions of the CFG which must be maintained, which again complicates later optimizations and so forth.

Basically: the reason to contemplate adding statements is to ensure that other optimizations *don't* have to think about this. These statements essentially are the same as DROP, but they are not terminators (and hence are only needed for variables that are Copy).

**Comment by nikomatsakis** [comment]
@nagisa 

> I don’t think I have much more outstanding comments other than my dislike about needlessly making exit_scope recursive again.

Heh. Recursion seemed pretty natural, but I guess I can rewrite it into a loop easily enough.

**Comment by nikomatsakis** [comment]
@nagisa made it iterative.

File: src/librustc_mir/build/scope.rs [hunk 20]
```diff
@@ -456,21 +456,41 @@ impl<'a,'tcx> Builder<'a,'tcx> {
     /// See module comment for more details. None indicates there’s no
     /// cleanup to do at this point.
     pub fn diverge_cleanup(&mut self) -> Option<BasicBlock> {
-        if self.scopes.is_empty() {
+        if self.scopes.iter().all(|scope| scope.drops.is_empty() && scope.free.is_none()) {
             return None;
         }
+        assert!(!self.scopes.is_empty()); // or `all` above would be true
```
Comment (nagisa) [inline src/librustc_mir/build/scope.rs]: Is this assertion really necessary or, perhaps, it is misplaced and should go above the `if`?

**Comment by nagisa** [comment]
Commits up to 0769865 look okay to me. It seems you have more plans in mind for this PR, so just ping me once it is ready for final review.

File: src/librustc_mir/build/scope.rs (hunk 20 above)
Comment (nikomatsakis) [inline src/librustc_mir/build/scope.rs]: @nagisa it's more a comment than an assertion, I guess. Although I think that, in practice, the list of scopes cannot be empty here, so it could in principle go above the `if`

**Comment by nikomatsakis** [comment]
@nagisa meh, I think I may well just land what you reviewed, and we can iterate more in other PRs.

**Comment by nikomatsakis** [comment]
(@pnkfelix reviewed the visitor commit)
